src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -316,9 +316,16 @@ zkutil::ZooKeeperPtr StorageReplicatedMergeTree::getZooKeeperAndAssertNotReadonl
     /// There's a short period of time after connection loss when new session is created,
     /// but replication queue is not reinitialized. We must ensure that table is not readonly anymore
     /// before using new ZooKeeper session to write something (except maybe GET_PART) into replication log.
-    auto res = getZooKeeper();
-    assertNotReadonly();
-    return res;
+    return getZooKeeperIfNotReadonly(/* assert_if_readonly = */ true, /* assert_if_readonly_while_shutdown = */ true);
+}
+
+zkutil::ZooKeeperPtr StorageReplicatedMergeTree::getZooKeeperIfNotReadonly(bool assert_if_readonly, bool assert_if_readonly_while_shutdown) const
```

Comment:
I find this new API confusing and hard to follow, as not it can do 3 things:
* Return a ZooKeeperPtr
* Return a nullptr
* Throw an exception.

The fact that all usages need to be next to a comment explaining what happened kinds of makes the point clear.

Maybe it should be something along the lines of:
```
auto zookeeper = tryGetZooKeeper(/* throw_on_shutdown */ false, /* throw_on_readonly */ true);
```

WDYT?

Reply:
Yes, I agree it was confusing. I tried to rework it the way you said but it still looked confusing because a table is readonly some time during the table's shutdown, however it is readonly also during the table's initialization and also when disconnected to ZooKeeper. I tried to write the required conditions in `tryGetZooKeeper()` but they seemed complicated.

So in order to not get even a more complicated code than it is now I finally decided to explicitly check the required conditions where they're needed without putting them into some tricky function.